### PR TITLE
[Typechecker] Fix duplicate diagnostics when using attributes

### PR DESCRIPTION
**Pull request by theblixguy**
We were not invalidating the `TypeRepr` after diagnosing in `resolveAttributedType`, which meant that trying to resolving the type for the same `TypeRepr` would result in the same diagnostic multiple times.

It seems to affect just `@autoclosure` and `@escaping` in some cases (and `@noescape` which is not supported anymore), but I've applied the invalidation in all cases to future proof this.

Resolves SR-9178

### CODE REVIEW

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Review by xedin — COMMENTED** [review]
So seeing a pattern here - should we add a special `diagnose` method to `TypeResolver` just like we have for `FailureDiagnostic` which would also take `repr` and invalidate it internally?

**Comment by theblixguy** [comment]
@xedin Sure thing, I have done that now!

File: lib/Sema/TypeCheckType.cpp [hunk 1]
```diff
@@ -1778,6 +1778,13 @@ namespace {
       return diags.diagnose(std::forward<ArgTypes>(Args)...);
     }
 
+    template <typename... ArgTypes>
+    InFlightDiagnostic diagnose(TypeRepr *repr, ArgTypes &&... Args) const {
```
Comment (xedin) [inline lib/Sema/TypeCheckType.cpp]: One more nit - maybe it should be called `invalidateAndDiagnose` or `diagnoseAndInvalidate` or `diagnoseInvalidRepr` or just `diagnoseInvalid` for clarity?

Comment (theblixguy) [reply lib/Sema/TypeCheckType.cpp]: I like `diagnoseInvalid` since the first argument is a `TypeRepr`

Comment (xedin) [reply lib/Sema/TypeCheckType.cpp]: Sure, let's go with that!

Comment (theblixguy) [reply lib/Sema/TypeCheckType.cpp]: Done!

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Review by xedin — APPROVED** [review]
LGTM!

**Comment by xedin** [comment]
I have added @rintaro as a reviewer as well since he spend some time on type resolution as well.

**Review by rintaro — APPROVED** [review]
No objection to this. Thanks!

**Pull request merged**